### DISCUSSION TITLE
caching_sha2_password vs mysql_native_password

### DISCUSSION THREAD

**Discussion by dorkyboi** [laravel/framework#51844] [General] [answered]
MySQL 8+ uses `caching_sha2_password` for storing user passwords, and has not only disabled `mysql_native_password` by default but also made it deprecated.
In PHP, support for `caching_sha2_password` has been added since version 7.4.4.

Laravel is still stuck using `mysql_native_password`, however Laravel 11 requires PHP 8.2+.
It appears that the switch from `mysql_native_password` to `caching_sha2_password` is long overdue.
WDYT?

**Comment by ghost** [comment]
Hi. The transition would mean regenerating all passwords(for their hashes) from db?

**Reply by dorkyboi** [reply]
Might be wrong here, but here's what I think:

Users of the website (e.g. in the `users` table) should be fine, as their passwords are stored as string in a varchar type column and are generated/manipulated by Laravel itself and its Hash facade.

Users of MySQL database (whose credentials one puts into `.env` file) will, indeed, have to have their passwords rehashed.

**Reply by ghost** [reply]
But how can laravel rehash without having the plain text password?

UPSATE or this is affecting only mysql users? If so then laravel is not affected at all.

**Reply by dorkyboi** [reply]
Yes, I believe it is just the MySQL users' passwords.
Laravel projects should not be affected in any way.

**Reply by ghost** [reply]
@dorkyboi then why "Laravel is still stuck using mysql_native_password" ?

**Reply by dorkyboi** [reply]
That's what I'm wondering.
I've had to setup a Linux server recently and found out that MySQL 8+ have `mysql_native_password` plugin disabled by default.
To enable it, I had to find the MySQL server config file and add a line that was not even in the file before (I had to type it down myself). I believe this shows just how outdated `mysql_native_password` is.

**Reply by ghost** [reply]
I just looked in my laravel. The password and token are generated in php and stored as strings in DB

**Reply by dorkyboi** [reply]
Yes, that's precisely what I said in my very first reply. This means that data of Laravel projects is safe.
The issue is in how Laravel connects to the database.

**Reply by ghost** [reply]
The password from env is plain text. So it is only mysql's business how it stores the user hashed passwords.

**Reply by dorkyboi** [reply]
Well, you said it yourself - the password from .env has no bearing on how passwords are stored in `users` table.
Laravel uses it to connect to the database, so that later Laravel could `INSERT INTO users VALUES (...)`.
The connection is what's failing with a clean MySQL install - Laravel tries to log in to the database using `mysql_native_password`.

**Reply by ghost** [reply]
I think it is from the pdo-mysql extension of php if the connection fails.
https://stackoverflow.com/questions/50026939/php-mysqli-connect-authentication-method-unknown-to-the-client-caching-sha2-pa

As of PHP 7.4, this is no longer an issue. Support for caching_sha2 authentication method has been added to mysqlnd.

**Reply by dorkyboi** [reply]
I could not find any mentions of `mysql_native_password` or `caching_sha2_password` by searching Laravel project structure, so I think you are correct.
But this is what weirds me out - I've installed PHP 8.2 (it is the only version installed), I've then added php-mysql - shouldn't they work via `caching_sha2_password` out-of-the-box?

**Reply by ghost** [reply]
I have docker-php-ext-install mysqli  && docker-php-ext-install pdo && docker-php-ext-install pdo_mysql

And in that link that guy talks about mysqlnd: https://dev.mysql.com/downloads/connector/php-mysqlnd/ (https://stackoverflow.com/a/50027581)

**Reply by dorkyboi** [reply]
Thank you for the links, I'll have a look

**Reply by dorkyboi** [reply, marked as answer]
It looks like the issue was my own doing.

I have chosen `php-mysql` as a driver due to the lack of knowledge.
I should've installed `php-mysqlnd` as it is, apparently, a better alternative to `php-mysql`, and it is also its drop-in replacement.

So I have removed `php-mysql` and installed `php-mysqlnd`, set the authentication plugin back to `caching_sha2_password`.
Everything works fine so far.

Thank you, @marius-mcp, for helping me to figure this out. ❤️
